### PR TITLE
start trying to make our error handling a bit more generic

### PR DESCRIPTION
**Pull request by reaperhulk**
This adds a some helpers for "friendly" openssl error messages and removes error matching in three places.

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -1473,18 +1476,7 @@ def elliptic_curve_supported(self, curve):
...
-                    self._lib.ERR_LIB_EC, self._lib.EC_R_UNKNOWN_GROUP
-                )
-                or
-                # This occurs in FIPS mode for unsupported curves on RHEL
-                errors[0]._lib_reason_match(
-                    self._lib.ERR_LIB_EC, self._lib.EC_R_NOT_A_NIST_PRIME
-                )
-            )
+            self._consume_errors()
             return False
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py:1480]: This is a supported check and ultimately it doesn't matter why group is null, so we should just return false after consuming the error stack.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -2398,19 +2390,7 @@ def derive_scrypt(self, key_material, salt, length, n, r, p):
             length,
         )
         if res != 1:
-            errors = self._consume_errors()
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py:2401]: This is a bit more debatable. We will error with MemoryError no matter what so it won't proceed unsafely, but it is hypothetically possible this wouldn't be due to OOM. We could hedge by attaching the error stack here like we do in DH

File: src/cryptography/hazmat/backends/openssl/dh.py [hunk 3]
```diff
@@ -149,8 +139,11 @@ def exchange(self, peer_public_key):
         )
 
         if res == -1:
-            errors = self._backend._consume_errors()
-            return _handle_dh_compute_key_error(errors, self._backend)
+            errors_with_text = self._backend._consume_errors_with_text()
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/dh.py:142]: We only know of an invalid public key here, but we want to value error no matter what and this attaches errors_with_text to the ValueError to provide more context.

**Comment by alex** [comment]
flake8 and libre are broken

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (alex) [reply src/cryptography/hazmat/backends/openssl/backend.py:2401]: Attaching the error stack feels safe in basically all cases.

**Pull request merged**